### ISSUE TITLE
[Bug]: Querynode terminated with log: failed to Deserialize index, cardinal inner error

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#30857]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: cardinal-milvus-io-2.3-ef086dc-20240222
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar   
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. collection `laion_stable_4` has 58m-768d+  data, and the schema is:
```
{'auto_id': False, 'description': '', 'fields': [{'name': 'id', 'description': '', 'type': <DataType.INT64: 5>, 'is_primary': True, 'auto_id': False}, {'name': 'float_vector', 'description': '', 'type': <DataType.FLOAT_VECTOR: 101>, 'params': {'dim': 768}}, {'name': 'int64_pk_5b', 'description': '', 'type': <DataType.INT64: 5>, 'is_partition_key': True}, {'name': 'varchar_caption', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}, {'name': 'varchar_NSFW', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}, {'name': 'float64_similarity', 'description': '', 'type': <DataType.FLOAT: 10>}, {'name': 'int64_width', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_height', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_original_width', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'int64_original_height', 'description': '', 'type': <DataType.INT64: 5>}, {'name': 'varchar_md5', 'description': '', 'type': <DataType.VARCHAR: 21>, 'params': {'max_length': 8192}}], 'enable_dynamic_field': True}
```
2. reload collection (64 segments) -> concurrent requests: insert + delete + search + query
![image](https://github.com/milvus-io/milvus/assets/27288593/6781b598-646f-4e6a-b00c-158d77879fd7)

4. One querynode of total 4 terminated 134 error with error logs:
(Since the cardinal is private, please get in touch with me for more detailed querynode terminated logs)
```
E20240226 16:13:52.003870   607 FileIo.cpp:25] [CARDINAL][FileReader][milvus] Failed to open file : /var/lib/milvus/data/querynode/index_files/447990444064979058/1/_mem.index.bin
E20240226 16:13:52.005385   607 cardinal.cc:368] [KNOWHERE][Deserialize][milvus] Cardinal Inner Exception: std::exception
I20240226 16:13:52.005625   607 time_recorder.cc:49] [KNOWHERE][PrintTimeRecord][milvus] Load index: done (2.135270 ms)
 => failed to Deserialize index, cardinal inner error
non-Go function
    pc=0x7f58fbc2003b
non-Go function
    pc=0x7f58fbbff858
non-Go function
    pc=0x7f58fba998d0
non-Go function
    pc=0x7f58fbaa537b
non-Go function
    pc=0x7f58fbaa4358
non-Go function
    pc=0x7f58fbaa4d10
non-Go function
    pc=0x7f58fbde1bfe
runtime.cgocall(0x4749090, 0xc001774cd0)
    /usr/local/go/src/runtime/cgocall.go:157 +0x5c fp=0xc001774ca8 sp=0xc001774c70 pc=0x1a627bc
github.com/milvus-io/milvus/internal/querynodev2/segments._Cfunc_DeleteSegment(0x7f58f68d1700)
    _cgo_gotypes.go:475 +0x45 fp=0xc001774cd0 sp=0xc001774ca8 pc=0x4522a45
github.com/milvus-io/milvus/internal/querynodev2/segments.(*LocalSegment).Release.func1(0xc00175b2d8?)
    /go/src/github.com/milvus-io/milvus/internal/querynodev2/segments/segment.go:1037 +0x3a fp=0xc001774d08 sp=0xc001774cd0 pc=0x453d07a
github.com/milvus-io/milvus/internal/querynodev2/segments.(*LocalSegment).Release(0xc00175b290)
    /go/src/github.com/milvus-io/milvus/internal/querynodev2/segments/segment.go:1037 +0xa6 fp=0xc001774f48 sp=0xc001774d08 pc=0x453c826
github.com/milvus-io/milvus/internal/querynodev2/segments.remove({0x5744620, 0xc00175b290})
    /go/src/github.com/milvus-io/milvus/internal/querynodev2/segments/manager.go:542 +0x42 fp=0xc001775010 sp=0xc001774f48 pc=0x452e5c2
github.com/milvus-io/milvus/internal/querynodev2/segments.(*segmentManager).Remove(0xc001620a80, 0x5158d57?, 0x3)
    /go/src/github.com/milvus-io/milvus/internal/querynodev2/segments/manager.go:447 +0x2d5 fp=0xc0017750b0 sp=0xc001775010 pc=0x452d5b5
```

### Milvus Log

- argo: [laion1b-test-new-1
](https://argo-workflows.zilliz.cc/archived-workflows/qa/cf8e6358-ff62-4855-a828-1dad200fe861?nodeId=laion1b-test-new-1)

- pods:
```
laion1b-test-2-etcd-0                                             1/1     Running             1 (46d ago)     74d     10.104.25.31    4am-node30   <none>           <none>
laion1b-test-2-etcd-1                                             1/1     Running             0               74d     10.104.30.94    4am-node38   <none>           <none>
laion1b-test-2-etcd-2                                             1/1     Running             0               74d     10.104.34.225   4am-node37   <none>           <none>
laion1b-test-2-milvus-datanode-7b7f99b8d4-g8v8q                   1/1     Running             0               20h     10.104.16.187   4am-node21   <none>           <none>
laion1b-test-2-milvus-datanode-7b7f99b8d4-t7lfp                   1/1     Running             0               20h     10.104.30.131   4am-node38   <none>           <none>
laion1b-test-2-milvus-indexnode-c8c8f4584-2kbqd                   1/1     Running             0               15h     10.104.14.112   4am-node18   <none>           <none>
laion1b-test-2-milvus-indexnode-c8c8f4584-d6q6m                   1/1     Running             0               15h     10.104.9.46     4am-node14   <none>           <none>
laion1b-test-2-milvus-indexnode-c8c8f4584-lg4k7                   1/1     Running             0               15h     10.104.34.47    4am-node37   <none>           <none>
laion1b-test-2-milvus-indexnode-c8c8f4584-q9hcx                   1/1     Running             0               15h     10.104.17.50    4am-node23   <none>           <none>
laion1b-test-2-milvus-indexnode-c8c8f4584-vtvx5                   1/1     Running             0               15h     10.104.29.240   4am-node35   <none>           <none>
laion1b-test-2-milvus-mixcoord-74b896d49d-ljz4l                   1/1     Running             0               20h     10.104.18.222   4am-node25   <none>           <none>
laion1b-test-2-milvus-proxy-5cdb5b7d6b-w5h29                      1/1     Running             0               20h     10.104.19.6     4am-node28   <none>           <none>
laion1b-test-2-milvus-querynode-0-7977c8fdbf-8pfz2                1/1     Running             0               15h     10.104.17.49    4am-node23   <none>           <none>
laion1b-test-2-milvus-querynode-0-7977c8fdbf-cb9mn                1/1     Running             0               15h     10.104.28.101   4am-node33   <none>           <none>
laion1b-test-2-milvus-querynode-0-7977c8fdbf-dd9hn                1/1     Running             0               15h     10.104.33.74    4am-node36   <none>           <none>
laion1b-test-2-milvus-querynode-0-7977c8fdbf-zfhhz                1/1     Running             1 (15h ago)     15h     10.104.32.10    4am-node39   <none>           <none>
laion1b-test-2-pulsar-bookie-0                                    1/1     Running             0               74d     10.104.33.107   4am-node36   <none>           <none>
laion1b-test-2-pulsar-bookie-1                                    1/1     Running             0               50d     10.104.18.240   4am-node25   <none>           <none>
laion1b-test-2-pulsar-bookie-2                                    1/1     Running             0               74d     10.104.25.32    4am-node30   <none>           <none>
laion1b-test-2-pulsar-broker-0                                    1/1     Running             0               68d     10.104.1.69     4am-node10   <none>           <none>
laion1b-test-2-pulsar-proxy-0                                     1/1     Running             0               74d     10.104.4.218    4am-node11   <none>           <none>
laion1b-test-2-pulsar-recovery-0                                  1/1     Running             0               74d     10.104.14.151   4am-node18   <none>           <none>
laion1b-test-2-pulsar-zookeeper-0                                 1/1     Running             0               74d     10.104.29.87    4am-node35   <none>           <none>
laion1b-test-2-pulsar-zookeeper-1                                 1/1     Running             0               74d     10.104.21.124   4am-node24   <none>           <none>
laion1b-test-2-pulsar-zookeeper-2                                 1/1     Running             0               74d     10.104.34.229   4am-node37   <none>           <none>
```

- core dump file:
`/tmp/cores/core-laion1b-test-2-milvus-querynode-0-7977c8fdbf-zfhhz-milvus-8-1708964037` of 4am-node39

**Comment by ThreadDao** [comment]
Perhaps it is because the `dataCoord.channel.watchTimeoutInterval` configuration is modified and the milvus is restarted. I mean when the qn restarts it looks like the tests haven't started yet

**Comment by foxspy** [comment]
![image](https://github.com/milvus-io/milvus/assets/11503321/af803fc0-4208-4f77-9f0c-833235d489d7)
The root cause seems that the concurrency bug between release and load. The qn release a segments while the index engine loading the index from file concurrently. And the index engine throwing an exception is as expected, which will not cause the qn coredump, and the actual cause of the coredump is the release operation.

**Comment by chyezh** [comment]
Index `447990444064979058` belongs to Segment `447990444064723266`.

Node `` start to release segment while new load request is incoming.

```

[2024/02/26 16:13:34.948 +00:00] [INFO] [querynodev2/services.go:595] ["start to release segments"] [traceID=06cd8e504609e669a530c57535e33631] [collectionID=447902879639453431] [shard=laion1b-test-2-rootcoord-dml_5_447902879639453431v1] [segmentIDs="[447990444064723266]"] [currentNodeID=1681]

[2024/02/26 16:13:38.996 +00:00] [INFO] [querynodev2/services.go:433] ["received load segments request"] [traceID=7c9ecfdda5b9a070dc760feb81f2bf64] [collectionID=447902879639453431] [partitionID=447902879639453437] [shard=laion1b-test-2-rootcoord-dml_5_447902879639453431v1] [segmentID=447990444064723266] [currentNodeID=1681] [version=1708964018906870595] [needTransfer=false] [loadScope=Full]
```

Load repeat segment is checked by SegmentManager.
```
...
		if len(loader.manager.Segment.GetBy(WithType(segmentType), WithID(segment.GetSegmentID()))) == 0 &&
			!loader.loadingSegments.Contain(segment.GetSegmentID()) {
...
```

Release segment is remove the segment from SegmentManager then release the memory.
```
	case querypb.DataScope_Historical:
		sealed = mgr.removeSegmentWithType(SegmentTypeSealed, segmentID)
		if sealed != nil {
			removeSealed = 1
		}

	mgr.updateMetric()
	mgr.mu.Unlock()

	if sealed != nil {
		remove(sealed)
	}
```

Concurrent load and release happens.

**Comment by chyezh** [comment]
Short-term fix: Implement mutual exclusivity between Release and Load on QN; 
Long-term, it is necessary to implement lifecycle controls such as Loading, Loaded, Release state of Segment on QueryCoord.

**Comment by congqixia** [comment]
@chyezh 
Loading segment will not be released in segment manager. In my opinion, concurrent load&release shall not happen for same segment. Could you please explain the detail how it went?

**Comment by chyezh** [comment]
> @chyezh Loading segment will not be released in segment manager. In my opinion, concurrent load&release shall not happen for same segment. Could you please explain the detail how it went?

Load is triggered when segment is releasing, but not release is triggered when segment is loading.

The release segment operation is divided into two steps on query node.
1. Removing the segment from the segmentManager (after this, the SegmentLoader is allowed to reload this segment),
3. Releasing the actual segment.

**Comment by congqixia** [comment]
> > @chyezh Loading segment will not be released in segment manager. In my opinion, concurrent load&release shall not happen for same segment. Could you please explain the detail how it went?
> 
> Load is triggered when segment is releasing, but not release is triggered when segment is loading.
> 
> The release segment operation is divided into two steps on query node.
> 
>     1. Removing the segment from the segmentManager (after this, the SegmentLoader is allowed to reload this segment),
> 
>     2. Releasing the actual segment.

@chyezh got it, thanks!

**Comment by congqixia** [comment]
After some offline discussion, the final solution shall be separating the disk resource for different segment life-cycle.

One more thing, it's looks weird that a segment is released than loaded back. Maybe the segment was bouncing between querynode?

**Comment by chyezh** [comment]
> After some offline discussion, the final solution shall be separating the disk resource for different segment life-cycle.
> 
> One more thing, it's looks weird that a segment is released than loaded back. Maybe the segment was bouncing between querynode?

- segment is released on QN for collection released.
- segment is reloaded for segment checker(lack of segment), updated by `Distribution`?

```
[2024/02/26 16:13:34.498 +00:00] [INFO] [task/scheduler.go:269] ["task added"] [task="[id=1708948067586] [type=Reduce] [source=segment_checker] [reason=collection released] [collectionID=447902879639453431] [replicaID=-1] [priority=Normal] [actionsCount=1] [actions={[type=Reduce][node=1681][streaming=false]}] [segmentID=447990444064723266]"]

[2024/02/26 16:13:38.501 +00:00] [INFO] [task/scheduler.go:269] ["task added"] [task="[id=1708948067608] [type=Grow] [source=segment_checker] [reason=lacks of segment] [collectionID=447902879639453431] [replicaID=447990457955778562] [priority=Normal] [actionsCount=1] [actions={[type=Grow][node=1681][streaming=false]}] [segmentID=447990444064723266]"]

```

**Comment by chyezh** [comment]
Release then load collection.
Or concurrent release and load collection can reproduce it.

**Comment by ThreadDao** [comment]
reproduce...
- image: cardinal-milvus-io-2.3-3c90475-20240311
- queryNode `laion1b-test-2-milvus-querynode-1-86cfff6f5d-7b2lv` terminated with 134 error at  2024-03-12 16:02:40.814(UTC)
![image](https://github.com/milvus-io/milvus/assets/27288593/d65f0631-1546-458c-8edd-3645d60aa485)

**Comment by ThreadDao** [comment]
Short-term fix: Implement mutual exclusivity between Release and Load on QN;

**Comment by chyezh** [comment]
should be fixed at 2.4.5, please verify it.
